pkg/cli/tail.go

```diff
@@ -0,0 +1,254 @@
+		Aliases: []string{"d"},
+		Target:  &c.flagDuration,
+		Example: "2h",
+		Default: 24 * time.Hour,
+		Usage:   `How far back to search for log entries, default is 24 hours`,
+	})
+
+	f.StringVar(&cli.StringVar{
+		Name:    "query",
+		Target:  &c.flagCustomQuery,
+		Example: `resource.type = "gae_app" AND severity = ERROR`,
+		Usage: `Optional custom log queries, see more on ` +
+			`https://cloud.google.com/logging/docs/view/logging-query-language`,
+	})
+
+	f.BoolVar(&cli.BoolVar{
+		Name:    "remove-lumberjack-log-type",
+		Target:  &c.flagRemoveLumberjackLogType,
+		Default: false,
+		Usage:   `Turn on to remove lumberjack log type log filter`,
+	})
+
+	f.BoolVar(&cli.BoolVar{
+		Name:    "validate-with-additional-check",
+		Target:  &c.flagValidateWithAdditionalCheck,
+		Default: false,
+		Usage: `Turn on for lumberjack log validation with additional ` +
+			`lumberjack specific checks on log labels.`,
+	})
+
+	return set
+}
+
+func (c *TailCommand) Run(ctx context.Context, args []string) error {
+	f := c.Flags()
+	if err := f.Parse(args); err != nil {
+		return fmt.Errorf("failed to parse flags: %w", err)
+	}
+	args = f.Args()
+	if len(args) > 0 {
+		return fmt.Errorf("unexpected arguments: %q", args)
+	}
+
+	if c.flagResource == "" {
+		return fmt.Errorf("resource is required")
+	}
+
+	// Request with negative and greater than 1000 (log count limit) is rejected.
+	if c.flagMaxNum <= 0 || c.flagMaxNum > 1000 {
+		return fmt.Errorf("max number must be greater than 0 and less than 1000")
+	}
+
+	// Tail logs.
+	ls, err := c.tail(ctx)
+	if err != nil {
+		return err
+	}
+	if len(ls) == 0 {
+		c.Outf("Log not found")
+		return nil
+	}
+
+	// Output results.
+	var extra []validation.Validator
+	if c.flagValidateWithAdditionalCheck {
+		extra = append(extra, validation.ValidateLabels)
+	}
+	var retErr error
+	for _, l := range ls {
+		js, err := protojson.Marshal(l)
+		if err != nil {
+			retErr = errors.Join(retErr, fmt.Errorf("failed to marshal log to json (InsertId: %q): %w", l.InsertId, err))
+			continue
+		}
+
+		// Output log entry in JSON format if validation is not enabled.
+		if !c.flagValidate && !c.flagValidateWithAdditionalCheck {
+			c.Outf(string(js))
+			continue
+		}
+
+		// Output validation result if validation is enabled.
+		if err := validation.Validate(string(js), extra...); err != nil {
+			retErr = errors.Join(retErr, fmt.Errorf("failed to validate log (InsertId: %q): %w", l.InsertId, err))
+		} else {
+			c.Outf("Successfully validated log (InsertId: %q)", l.InsertId)
+		}
+	}
+
+	return retErr
+}
+
+func (c *TailCommand) tail(ctx context.Context) ([]*loggingpb.LogEntry, error) {
+	var p logPuller
+	if c.testPuller != nil {
+		p = c.testPuller
+	} else {
+		logClient, err := logging.NewClient(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create logging client: %w", err)
+		}
+		p = cloudlogging.NewPuller(ctx, logClient, c.flagResource)
+	}
+
+	ls, err := p.Pull(ctx, c.getFilter(), c.flagMaxNum)
```

Comment:
Should debug output the filter.

Reply:
I wonder do we really need to output filter, it should be obvious what filters are used (I can add the lumberjack log type to the cmd description)?

Reply:
Resolving, feel free to reopen.